tls/s2n_server_finished.c

```diff
@@ -45,14 +45,10 @@ S2N_RESULT s2n_finished_recv(struct s2n_connection *conn, uint8_t *local_verify_
     return S2N_RESULT_OK;
 }
 
-S2N_RESULT s2n_finished_send(struct s2n_connection *conn, uint8_t *seq_num, uint8_t *verify_data)
+S2N_RESULT s2n_finished_send(struct s2n_connection *conn, uint8_t *verify_data)
 {
     RESULT_ENSURE_REF(conn);
 
-    struct s2n_blob seq = { 0 };
-    RESULT_GUARD_POSIX(s2n_blob_init(&seq, seq_num, S2N_TLS_SEQUENCE_NUM_LEN));
-    RESULT_GUARD_POSIX(s2n_blob_zero(&seq));
-
```

Comment:
What about TLS1.3? I'm guessing we zero elsewhere / when we change keys? I'm pretty sure I wrote that code, but double-check.

And where is the TLS1.2 zeroing happening now, if anywhere? There are pretty strict rules about the sequence number being zero on encryption key change, so we should probably still do it somewhere even if we expect it to already be zero.

Reply:
So, in TLS1.2 secure->sequence number is zeroed when you receive the peer's CCS message. See: https://github.com/aws/s2n-tls/blob/main/tls/s2n_change_cipher_spec.c#L49.

In TLS1.3 we zero in two places, where we do a key update: https://github.com/aws/s2n-tls/blob/main/tls/s2n_tls13_handshake.c#L196 and where we do a key change: https://github.com/aws/s2n-tls/blob/main/tls/s2n_tls13_key_schedule.c#L50
So TLS1.3 code shouldn't be affected by this at all.

Reply:
Oh true. We don't zero our own sequence number. I'm relying on the fact that this is the first time conn->secure is used, so the sequence number is already zero in a regular handshake. I mean, if we wanted to make the least impact code-wise, we could gate this zeroizing based on whether or not NPN was negotiated. But then it doesn't really make sense for this to be a separate PR from #3545 .

Reply:
Should we zeroize when we send our CCS message?

Reply:
Ah, but that wouldn't cover the potential seq_num++ from actually sending the CCS record

Reply:
I think your issue is bigger than just zeroize. You also need to change where we enable encryption, don't you? https://github.com/aws/s2n-tls/blob/main/tls/s2n_server_finished.c#L81 How were you handling that with npn messages?

Because we should probably zeroize whenever we change to conn->secure.

Reply:
Hmmm this is a hard one. There's three 'changes' to the current handshake messages that need to happen to be able to negotiate NPN. 
1. Stop zeroizing the sequence number when you read a Finished message
2. Calculate the verify data later than receiving the CCS message. 
3. Switch to the secure crypto parameters earlier. 

I can move all of these to this PR, but it's harder to see why these changes are necessary without the self-talk test to back it up.

Reply:
Oh your updated comment makes it clearer. 2 and 3 can be handled by just repeated the action, not by necessarily moving it. So 1 is the only really tough one.